### DISCUSSION TITLE
Problem with setting plantit for reverse proxy

### DISCUSSION THREAD

**Discussion by lockheed** [MDeLuise/plant-it#151] [Q&A]
I am using caddy and the docker-compose for treafik is not clear to me.

Here's my docker-compose with external db.

> `version: "3"
> name: plant-it
> services:
>   server:
>     image: msdeluise/plant-it-server:latest
>     container_name: plant-it
>     environment:
>       MYSQL_HOST: mariadb
>       MYSQL_PORT: 3306
>       MYSQL_USERNAME: plantiter
>       MYSQL_PSW: pass
>       MYSQL_DATABASE: plantit
>       TZ: Europe/Warsaw
>       JWT_SECRET: plantpass
>       JWT_EXP: 1
>       USERS_LIMIT: -1
>       UPLOAD_DIR: /upload-dir
>       API_PORT: 8080
>       TREFLE_KEY: null
>       LOG_LEVEL: DEBUG
>       ALLOWED_ORIGINS: "*"
>       CACHE_TTL: 86400
>       CACHE_HOST: cache
>       CACHE_PORT: 6379
>       SSL_ENABLED: false
>       CERTIFICATE_PATH: /certificates/
>     volumes:
>       - /z/services/plantit/upload:/upload-dir
>       - /z/services/plantit/certs:/certificates
>     networks:
>       - frontend
>       - db-net
>     labels:
>       caddy: plantit.mydomain.com
>       caddy.reverse_proxy: "* {{upstreams 3000}}"
>     restart: unless-stopped
>   cache:
>     image: redis:7.2.1
>     container_name: plant-it-cache
>     restart: always
>     networks:
>       - db-net
> networks:
>   frontend:
>     external: true
>   db-net:
>     external: true
> 

Then I get to the homepage, but the first step there is to set the URL and it doesn't accept my domain.

Do I need TWO public domains to host PlantIt? One for 3000 and the other one for 8080? Why would API need a separate public domain? Unfortunately, your documentation says nothing about it.

**Comment by MDeLuise** [comment]
Hi @lockheed, thanks for opening the issue.
I'm seeing that some users have problems with nginx/traefik, I plant to create a local dummy environment to inspect the behavior more closely.

In general, the URL you should insert in the server URL configuration is the same as the one used to access the Swagger UI console. So, if you're able to reach the Swagger UI, the application should theoretically be able to connect to the server.

Could you please confirm whether you're able to connect to Swagger at `<url_you_try_to_use>:<port>/api/swagger-ui/index.html`? The default port in the docker-compose file is 8080, but if you've changed the port binding, please use that value instead.

If everything works fine, you should see something like this
<img width="30%" alt="Screenshot 2024-05-08 at 16 02 20" src="https://github.com/MDeLuise/plant-it/assets/66636702/af89ca21-8f74-41f3-babe-6ec2ff804d2e">

**Reply by szszl0** [reply]
i have the same problem while using nginx proxy manager. I can connect to swagger but still when trying to use my https://plant.mydomain.xxx i get the "cannot connect to server error". Any solutions? Everything works in android app and the raw http://adress.xxx:3000...

**Reply by njack21** [reply]
+1 - I can't get this working properly behind NPM.

Note: I have both plant-it and NPM running in separate LXCs on a Proxmox node.

If I go to https://plant.mydomain.com, I am prompted with "let's make magic happen" and an field to enter my server URL. Every URL I enter returns the "cannot connect to server error" Maybe that is where magic is supposed to happen. If so, the magic is broken.

If I go to https://plant.mydomain.com:8080/api/swagger-ui/index.html, I am able to enter a server URL of http://192.x.x.x:8080 and get into the app. However, https://plant.mydomain.com never works, I have to specify the full https://plant.mydomain.com:8080/api/swagger-ui/index.html in my URL, regardless of my NPM configuration.

Does anyone have a working config and can spare the magic part of the configuration?

**Comment by lockheed** [comment]
The port part is what doesn't make sense to me. Reverse proxy by nature only forward one port, and it is port 443 (HTTPS) and nothing else, so how could it  access 8080? It cannot.

This is the only web app I know of that requires more than one port to get into.

**Reply by martinbjeldbak** [reply]
You can certainly run a reverse proxy fronting multiple ports. For example, here is my working Caddyfile


```
*.local.martinbjeldbak.com {
	@plantit host plant-it.local.martinbjeldbak.com
	handle @plantit {
		reverse_proxy plant-it:3000
	}

	@plantitapi host plant-it-api.local.martinbjeldbak.com
	handle @plantitapi {
		reverse_proxy plant-it:8080
	}

	# Fallback for otherwise unhandled domains
	handle {
		abort
	}
}
```

where `plant-it` is the service name of the container running in the docker network. In your case, it'd be `server`.

If curious, you can see full implementation in my repository here: https://github.com/martinbjeldbak/self-hosted

**Comment by mnfctrdinitaly** [comment]
I'm having the same issues... i've deployed behind nginx and using authelia to authenticate. (Although all the troubleshooting has been set to bypass authelia.)

I can access the welcome page using the subdomain "plants.example.tld" but when I try to use that to access the api/backend (using the suggest url `host` : `port` `/api/swagger-ui/index.html` i always get this error: 

>Error response
>Error code: 404
>
>Message: File not found.
>
>Error code explanation: 404 - Nothing matches the given URI.

The nginx conf file has been tried in two configurations:
Config 1
>`subdomain.example.tld` proxy to `plant-it:3000`
>`subdomain2.axample.tld` proxy to `plant-it:8080`

Config 2
>`subdomain.example.tld/` proxy to `plant-it:3000`
>`subdomain.example.tld/api` proxy to `plant-it:8080`

Neither configuration works...

Please help?

**Reply by jmrplens** [reply]
For people who have problems with Nginx, when there are different ports must be configured different upstream, here I leave the configuration that works correctly for a docker installation on the same server as Nginx:

```bash
upstream plantitfrontend {
    server localhost:3000 fail_timeout=0; # Change port if you changed in docker.
}

upstream plantitbackend {
    server localhost:8080 fail_timeout=0; # Change port if you changed in docker.
}

# Upgrade WebSocket if requested, otherwise use keepalive
map $http_upgrade $connection_upgrade_keepalive {
    default upgrade;
    ''      '';
}

server {
    server_name plantit.domain.uri;

    # Listen Unsecure
    listen 80;
    listen [::]:80;
    
    # Redirect to https
    if ($scheme = "http") {
        return 301 https://$host$request_uri;
    }

    # Listen Secure Protocols
    listen 443 ssl;
    listen [::]:443 ssl;
    listen 443 quic;
    listen [::]:443 quic;
    http2 on;
    http3 on;
    quic_retry on;
    add_header Alt-Svc 'h3=":443"; ma=86400';
    
    # HSTS settings
    add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;
    add_header X-Frame-Options DENY;
    add_header X-Content-Type-Options nosniff;

    # CERTS
    ssl_certificate /path/to/cert.pem;
    ssl_certificate_key /path/to/key.pem;

    ssl_session_timeout 1d;
    ssl_session_cache shared:MozSSL:10m; 
    ssl_session_tickets off;

    ssl_protocols TLSv1.2 TLSv1.3;
    ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305;
    ssl_prefer_server_ciphers on;

    # Other Security settings
    add_header Referrer-Policy "no-referrer" always;
    add_header X-Content-Type-Options "nosniff" always;
    add_header X-Frame-Options "SAMEORIGIN" always;
    add_header X-Permitted-Cross-Domain-Policies "none" always;
    add_header X-XSS-Protection "1; mode=block" always;
    add_header Permissions-Policy "interest-cohort=()" always;
    
    proxy_buffering off;
    
    # Redirect to Plant-it REST API
    location  ^~ /api {
        proxy_pass http://plantitbackend; # Don't use https here
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade_keepalive;
        add_header 'Content-Security-Policy' 'upgrade-insecure-requests';
    }

    # Main frontend
    location / {
        proxy_pass http://plantitfrontend; # Don't use https here
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade_keepalive;
        add_header 'Content-Security-Policy' 'upgrade-insecure-requests';
    }

}
```

I hope it helps ;)

**Reply by Dvalin21** [reply]
Can you give more explanation please?  I'm not sure what Im suppose to do with this

**Reply by angel-torres** [reply]
I'm using nginx-proxy-manager with authelia. This is what I have in the Advanced tab for the plants.* proxy host. How would I apply the  above updates to my configuration?

```
location /authelia {
    internal;
    set $upstream_authelia http://xxx.xxx.x.xxx:xxxx/api/verify; #ADD YOUR IP AND PORT OF AUTHELIA
    proxy_pass_request_body off;
    proxy_pass $upstream_authelia;
    proxy_set_header Content-Length "";

    # Timeout if the real server is dead
    proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;
    client_body_buffer_size 128k;
    proxy_set_header Host $host;
    proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $http_host;
    proxy_set_header X-Forwarded-Uri $request_uri;
    proxy_set_header X-Forwarded-Ssl on;
    proxy_redirect  http://  $scheme://;
    proxy_http_version 1.1;
    proxy_set_header Connection "";
    proxy_cache_bypass $cookie_session;
    proxy_no_cache $cookie_session;
    proxy_buffers 4 32k;

    send_timeout 5m;
    proxy_read_timeout 240;
    proxy_send_timeout 240;
    proxy_connect_timeout 240;
}

location / {
    set $upstream_plants http://plant-it:3000;
    proxy_pass $upstream_plants;

    auth_request /authelia;
    auth_request_set $target_url $scheme://$http_host$request_uri;
    auth_request_set $user $upstream_http_remote_user;
    auth_request_set $groups $upstream_http_remote_groups;
    proxy_set_header Remote-User $user;
    proxy_set_header Remote-Groups $groups;
    error_page 401 =302 https://auth.mydomain.com/?rd=$target_url;

    client_body_buffer_size 128k;

    proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;

    send_timeout 5m;
    proxy_read_timeout 360;
    proxy_send_timeout 360;
    proxy_connect_timeout 360;

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $http_host;
    proxy_set_header X-Forwarded-Uri $request_uri;
    proxy_set_header X-Forwarded-Ssl on;
    proxy_redirect  http://  $scheme://;
    proxy_http_version 1.1;
    proxy_set_header Connection "";
    proxy_cache_bypass $cookie_session;
    proxy_no_cache $cookie_session;
    proxy_buffers 64 256k;

    set_real_ip_from xxx.xxx.x.0/24;
    real_ip_header X-Forwarded-For;
    real_ip_recursive on;

}
```

**Reply by mr8tom** [reply]
Config from @jmrplens **works**. I had to put the same URL into the step to set the URL like the Domain. No /api
Example: https://plant-it.example.com -> URL Backend = https://plant-it.example.com

**Comment by isaaclw** [comment]
My understanding of docker was that you were supposed to be able to deploy without any configurations and get an environment set up. Is that not the case?

I don't mind rifling through the weeds to get this working, but then it wouldn't be docker.

Maybe developers who are working on this project haven't been testing the recent docker containers?

**Reply by njack21** [reply]
I'm not convinced they've testing anything behinda reverse proxy.

**Reply by lukyjay** [reply]
> My understanding of docker was that you were supposed to be able to deploy without any configurations and get an environment set up. Is that not the case?

The docker container works. You're trying to connect two docker containers (Plant-it and a reverse proxy). There are no guarantees that this service will work with your setup.

> I don't mind rifling through the weeds to get this working, but then it wouldn't be docker.
> 
> Maybe developers who are working on this project haven't been testing the recent docker containers?

This is the benefit of free open source software. You can debug and develop this yourself, otherwise try funding the developer to do it for you. Why would they dedicate hours of their spare time to work on a feature that they themselves don't need?

**Reply by isaaclw** [reply]
> > My understanding of docker was that you were supposed to be able to deploy without any configurations and get an environment set up. Is that not the case?
> 
> The docker container works. You're trying to connect two docker containers (Plant-it and a reverse proxy). There are no guarantees that this service will work with your setup.

I'm actually not doing anything complicated, I'm doing exactly what the documentation says (it says almost nothing, so I guess I shouldn't be surprised).

I posted here (https://github.com/MDeLuise/plant-it/discussions/456) and got zero replies, so I came here in case there was anything remotely useful.

It does seem like this project is dead, unfortunately :(

**Reply by njack21** [reply]
> The docker container works. You're trying to connect two docker containers (Plant-it and a reverse proxy). There are no guarantees that this service will work with your setup.

This is a pretty typical setup that works with literally every other service I host.

**Reply by aadityajshah** [reply]
I got it to work with my reverseproxy by adding the trailing '/' for the backend. Maybe try that?

**Reply by angel-torres** [reply]
This is how I got it to work with Nginx-Proxy-Manager and Authelia. I have NPM and Authelia on a network called "proxy" and Plant-it on one called "services". You don't have to do it this way. It's just how I have mine set up.

**DOCKER-COMPOSE.YML**
Here is my Plant-it docker-compose.yml. I only expose 8080 and 3000. I didn't use the "ports" because my Docker is running in WSL2 on a Windows host. I'm already using the default ports for other containers but NPM only needs the container name and the default ports.
```
services:
  plant-it:
    image: msdeluise/plant-it-server:latest
    container_name: plant-it
    env_file: server.env
    depends_on:
      - plant-it-db   # same name as the service below
      - plant-it-cache   # same name as the service below
    restart: unless-stopped
    volumes:
      - "./upload-dir:/upload-dir"
    expose:
      - 8080
      - 3000
    networks:
      - services
      - proxy   # has to be on both networks so NPM can see it

  plant-it-db:
    image: mysql:8.0
    container_name: plant-it-db
    restart: always
    env_file: server.env
    command: --lower_case_table_names=2
    volumes:
      - "./db:/var/lib/mysql"
    networks:
      - services

  plant-it-cache:
    image: redis:7.2.1
    container_name: plant-it-cache
    restart: always
    networks:
      - services
```

**SERVER.ENV (or just .env):**
I didn't change much in the .env file. Made sure CACHE_HOST=plant-it-cache (same name as the container/service), I went to floracodex.com to get the FLORACODEX_KEY. I have a Yahoo account that I use for the SMTP info. With Yahoo, you have to use an app password as the SMTP_PASSWORD (same with Outlook).

**NGINX-PROXY-MANAGER:**

- Domain Names = plant-it.*
- Scheme = http
- Forward Hostname / IP = plant-it
- Forward Port = 3000
- Cache Assets, Block Common Exploits, and Websockets Support = All True
- SSL tab = I have my certificate selected and all items set to True.
- Advanced tab (maybe some of these translate to Traefik/Caddy?):
```
# AUTHELIA SECTION (NOT NEEDED IF YOU DON'T USE AUTHELIA)
location /authelia {
    internal;
    set $upstream_authelia http://authelia:####/api/verify;
    proxy_pass_request_body off;
    proxy_pass $upstream_authelia;
    proxy_set_header Content-Length "";

    # Timeout if the real server is dead
    proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;
    client_body_buffer_size 128k;
    proxy_set_header Host $host;
    proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $http_host;
    proxy_set_header X-Forwarded-Uri $request_uri;
    proxy_set_header X-Forwarded-Ssl on;
    proxy_redirect  http://  $scheme://;
    proxy_http_version 1.1;
    proxy_set_header Connection "";
    proxy_cache_bypass $cookie_session;
    proxy_no_cache $cookie_session;
    proxy_buffers 4 32k;

    send_timeout 5m;
    proxy_read_timeout 240;
    proxy_send_timeout 240;
    proxy_connect_timeout 240;
}

# THIS IS FOR THE PLANT-IT API
location ^~ /api {
    # PROXY SETTINGS
    set $upstream_plants http://plant-it:8080;  # USE THE API PORT
    proxy_pass $upstream_plants;

    # SECTION FOR AUTHELIA SETTINGS (NOT NEEDED IF YOU DON'T USE AUTHELIA)
    auth_request /authelia;
    auth_request_set $target_url $scheme://$http_host$request_uri;
    auth_request_set $user $upstream_http_remote_user;
    auth_request_set $groups $upstream_http_remote_groups;
    proxy_set_header Remote-User $user;
    proxy_set_header Remote-Groups $groups;
    error_page 401 =302 https://auth.YOURDOMAIN.COM/?rd=$target_url;  # CHANGE "YOURDOMAIN.COM" TO YOUR OWN

    # MORE PROXY SETTINGS
    client_body_buffer_size 128k;
    proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;
    send_timeout 5m;
    proxy_read_timeout 360;
    proxy_send_timeout 360;
    proxy_connect_timeout 360;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $http_host;
    proxy_set_header X-Forwarded-Uri $request_uri;
    proxy_set_header X-Forwarded-Ssl on;
    proxy_redirect  http://  $scheme://;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_cache_bypass $cookie_session;
    proxy_no_cache $cookie_session;
    proxy_buffers 64 256k;
    set_real_ip_from 192.168.1.0/16;  # SUBNET MASK OF YOUR INTERNAL NETWORK
    real_ip_header X-Forwarded-For;
    real_ip_recursive on;
}

# THIS IS FOR THE MAIN PLANT-IT PAGE
location / {
    # PROXY SETTINGS
    set $upstream_plants http://plant-it:3000;  # USE THE MAIN PORT
    proxy_pass $upstream_plants;

    # SECTION FOR AUTHELIA SETTINGS (NOT NEEDED IF YOU DON'T USE AUTHELIA)
    auth_request /authelia;
    auth_request_set $target_url $scheme://$http_host$request_uri;
    auth_request_set $user $upstream_http_remote_user;
    auth_request_set $groups $upstream_http_remote_groups;
    proxy_set_header Remote-User $user;
    proxy_set_header Remote-Groups $groups;
    error_page 401 =302 https://auth.YOURDOMAIN.COM/?rd=$target_url; # CHANGE "YOURDOMAIN.COM" TO YOUR OWN

    # MORE PROXY SETTINGS
    client_body_buffer_size 128k;
    proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;
    send_timeout 5m;
    proxy_read_timeout 360;
    proxy_send_timeout 360;
    proxy_connect_timeout 360;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $http_host;
    proxy_set_header X-Forwarded-Uri $request_uri;
    proxy_set_header X-Forwarded-Ssl on;
    proxy_redirect  http://  $scheme://;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_cache_bypass $cookie_session;
    proxy_no_cache $cookie_session;
    proxy_buffers 64 256k;
    set_real_ip_from 192.168.1.0/16;  # SUBNET MASK OF YOUR INTERNAL NETWORK
    real_ip_header X-Forwarded-For;
    real_ip_recursive on;
}
```
